managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java

```diff
@@ -249,7 +249,7 @@ private void asyncReadEntry0(ReadHandle lh, Position position, final ReadEntryCa
             manager.mlFactoryMBean.recordCacheHit(cachedEntry.getLength());
             callback.readEntryComplete(cachedEntry, ctx);
         } else {
-            lh.readAsync(position.getEntryId(), position.getEntryId()).thenAcceptAsync(
+            lh.readUnconfirmedAsync(position.getEntryId(), position.getEntryId()).thenAcceptAsync(
```

Comment:
Do we need to check the `position.getEntryId()` with LAC? Otherwise, it may return the unconfirmed entry.

Reply:
The BK client side (managed ledger) should be responsible to guarantee the `position` is confirmed.

Take this `asyncReadEntry0` method for example, the call stack is:

```
ManagedLedgerImpl#asyncReadEntry
RangeEntryCacheImpl#asyncReadEntry
RangeEntryCacheImpl#asyncReadEntry0
```

Here are all references of `asyncReadEntry`:

```java
    public CompletableFuture<Long> getEarliestMessagePublishTimeOfPos(Position pos) {
        /* ... */
        if (nextPos.compareTo(lastConfirmedEntry) > 0) {
            return CompletableFuture.completedFuture(-1L);
        }

        asyncReadEntry(nextPos, new ReadEntryCallback() {
```

```java
// ManagedCursorImpl#asyncGetNthEntry, in getPositionAfterN, `lastConfirmedEntry` is also checked
                Position positionAfterN = ledger.getPositionAfterN(markDeletePosition, n + deletedMessages,
                        PositionBound.startExcluded);
                ledger.asyncReadEntry(positionAfterN, callback, ctx);
```

I didn't look into the `ManagedCursorImpl#asyncReplayEntries` call, but the positions should come from the entries that have been consumed so that they should be confirmed now.

Reply:
In normal cases, the max position cannot be greater than `PersistentTopic#getLastPosition`:

```java
    public Position getLastPosition() {
        return ledger.getLastConfirmedEntry();
    }
```

see the `readMoreEntries` method call in dispatcher:

```java
                    cursor.asyncReadEntriesOrWait(messagesToRead,
                            bytesToRead, this, readEntriesCtx, topic.getMaxReadPosition());
```

Reply:
I added a compare with `lastConfirmedEntry` to fail when the parameters are wrong.